compiler/qsc_data_structures/src/namespaces.rs

```diff
@@ -0,0 +1,304 @@
+    fn from(value: usize) -> Self {
+        Self::new(value)
+    }
+}
+
+impl From<NamespaceId> for usize {
+    fn from(value: NamespaceId) -> Self {
+        value.0
+    }
+}
+
+impl From<&NamespaceId> for usize {
+    fn from(value: &NamespaceId) -> Self {
+        value.0
+    }
+}
+
+impl Deref for NamespaceId {
+    type Target = usize;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Display for NamespaceId {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Namespace {}", self.0)
+    }
+}
+
+/// A reference counted cell that supports interior mutability for namespace tree nodes.
+/// Interior mutability is required to update the tree when inserting new data structures.
+type NamespaceTreeCell = Rc<RefCell<NamespaceTreeNode>>;
+
+/// An entry in the memoization table for namespace ID lookups.
+type MemoEntry = (Vec<Rc<str>>, NamespaceTreeCell);
+
+/// The root of the data structure that represents the namespaces in a program.
+/// The tree is a trie (prefix tree) where each node is a namespace and the children are the sub-namespaces.
+/// For example, the namespace `Microsoft.Quantum.Canon` would be represented as a traversal from the root node:
+/// ```
+/// root
+/// └ Microsoft
+///   └ Quantum
+///     └ Canon
+/// ```
+/// This data structure is optimized for looking up namespace IDs by a given name. Looking up a namespace name by ID is
+/// less efficient, as it performs a breadth-first search. Because of this inefficiency, the results of this lookup are memoized.
+/// [`NamespaceTreeNode`]s are all stored in [`NamespaceTreeCell`]s, which are reference counted and support interior mutability for namespace
+/// insertions and clone-free lookups.
+#[derive(Debug, Clone)]
+pub struct NamespaceTreeRoot {
+    assigner: usize,
+    tree: NamespaceTreeCell,
+    memo: RefCell<FxHashMap<NamespaceId, MemoEntry>>,
+}
+
+impl NamespaceTreeRoot {
+    /// Create a new namespace tree root. The assigner is used to assign new namespace IDs.
+    #[must_use]
+    pub fn new_from_parts(assigner: usize, tree: NamespaceTreeNode) -> Self {
+        Self {
+            assigner,
+            tree: Rc::new(RefCell::new(tree)),
+            memo: RefCell::new(FxHashMap::default()),
+        }
+    }
+
+    /// Get the namespace tree field. This is the root of the namespace tree.
+    #[must_use]
+    pub fn tree(&self) -> NamespaceTreeCell {
+        self.tree.clone()
+    }
+
+    /// Insert a namespace into the tree. If the namespace already exists, return its ID.
+    pub fn insert_or_find_namespace(
+        &mut self,
+        ns: impl IntoIterator<Item = Rc<str>>,
+    ) -> NamespaceId {
+        self.tree
+            .borrow_mut()
+            .insert_or_find_namespace(ns.into_iter().peekable(), &mut self.assigner)
+            .expect("namespace creation should not fail")
+    }
+
+    /// Create a new namespace node with the given children.
+    pub fn new_namespace_node(
```

Comment:
This function creates a new node, but doesn't insert it anywhere into the tree, though it _does_ increment the assigner. This seems like dangerous behavior for a pub function.

Reply:
I'm not seeing this used anywhere. Might be better to not have this function.